File: backend/app/api/v1/routers/system.py
from datetime import datetime

from fastapi import APIRouter, Depends, Query, Request, status
from pydantic import Field
from sqlalchemy.orm import Session

from app.core.dependencies import get_current_active_user, get_db, require_permission
from app.models.system import SysUser
from app.schemas.common import PageResult
from app.schemas.system import (
    OrgDepartmentCreate,
    OrgDepartmentResponse,
    OrgDepartmentUpdate,
    SysAuditLogResponse,
    SysPermissionCreate,
    SysPermissionResponse,
    SysPermissionUpdate,
    SysRoleCreate,
    SysRoleDetail,
    SysRoleResponse,
    SysRoleUpdate,
    SysUserAdminResetPassword,
    SysUserChangePassword,
    SysUserCreate,
    SysUserDetail,
    SysUserResponse,
    SysUserUpdate,
)
from app.services.system_service import (
    AuditLogService,
    DepartmentService,
    PermissionService,
    RoleService,
    UserService,
)

router = APIRouter()
def _commit_write(db: Session) -> None:
    db.commit()


@router.get(
    "/departments",
    response_model=PageResult[OrgDepartmentResponse],
    summary="List departments",
    description="Filter by keyword and active status with pagination.",
)
def list_departments(
    keyword: str | None = Query(None, description="Department name or code keyword"),
    is_active: bool | None = Query(None, description="Filter by active status"),
    page: int = Query(1, ge=1, description="Page number"),
    size: int = Query(20, ge=1, le=500, description="Page size"),
    db: Session = Depends(get_db),
    _: SysUser = Depends(require_permission("/system/departments", "read")),
) -> PageResult[OrgDepartmentResponse]:
    return DepartmentService(db).list(keyword=keyword, is_active=is_active, page=page, size=size)


@router.post(
    "/departments",
    response_model=OrgDepartmentResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create department",
)
def create_department(
    payload: OrgDepartmentCreate,
    request: Request,
    db: Session = Depends(get_db),
    current_user: SysUser = Depends(require_permission("/system/departments", "write")),
) -> OrgDepartmentResponse:
    operator = str(current_user.id)
    result = DepartmentService(db).create(payload, operator)
    AuditLogService(db).write(
        user_id=current_user.id,
        username=current_user.username,
        action="CREATE",
        resource_type="OrgDepartment",
        resource_id=result.id,
        detail={"name": result.name, "code": result.code},
        ip_address=request.client.host if request.client else None,
    )
    _commit_write(db)
    return result


@router.get(
    "/departments/{dept_id}",
    response_model=OrgDepartmentResponse,
    summary="Get department",
)
def get_department(
    dept_id: int,
    db: Session = Depends(get_db),
    _: SysUser = Depends(require_permission("/system/departments", "read")),
) -> OrgDepartmentResponse:
    return DepartmentService(db).get(dept_id)


@router.put(
    "/departments/{dept_id}",
    response_model=OrgDepartmentResponse,
    summary="Update department",
)
def update_department(
    dept_id: int,
    payload: OrgDepartmentUpdate,
    request: Request,
    db: Session = Depends(get_db),
    current_user: SysUser = Depends(require_permission("/system/departments", "write")),
) -> OrgDepartmentResponse:
    operator = str(current_user.id)
    result = DepartmentService(db).update(dept_id, payload, operator)
    AuditLogService(db).write(
        user_id=current_user.id,
        username=current_user.username,
        action="UPDATE",
        resource_type="OrgDepartment",
        resource_id=dept_id,
        detail=payload.model_dump(exclude_unset=True),
        ip_address=request.client.host if request.client else None,
    )
    _commit_write(db)
    return result


@router.delete(
    "/departments/{dept_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete department",
)
def delete_department(
    dept_id: int,
    request: Request,
    db: Session = Depends(get_db),
    current_user: SysUser = Depends(require_permission("/system/departments", "delete")),
) -> None:
    operator = str(current_user.id)
    DepartmentService(db).delete(dept_id, operator)
    AuditLogService(db).write(
        user_id=current_user.id,
        username=current_user.username,
        action="DELETE",
        resource_type="OrgDepartment",
        resource_id=dept_id,
        ip_address=request.client.host if request.client else None,
    )
    _commit_write(db)


@router.get(
    "/permissions",
    response_model=PageResult[SysPermissionResponse],
    summary="List permissions",
)
def list_permissions(
    keyword: str | None = Query(None, description="Permission name, code, or resource path keyword"),
    action: str | None = Query(None, description="Action filter: read/write/delete/admin"),
    page: int = Query(1, ge=1),
    size: int = Query(20, ge=1, le=500),
    db: Session = Depends(get_db),
    _: SysUser = Depends(require_permission("/system/permissions", "read")),
) -> PageResult[SysPermissionResponse]:
    return PermissionService(db).list(keyword=keyword, action=action, page=page, size=size)


@router.post(
    "/permissions",
    response_model=SysPermissionResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create permission",
)
def create_permission(
    payload: SysPermissionCreate,
    request: Request,
    db: Session = Depends(get_db),
    current_user: SysUser = Depends(require_permission("/system/permissions", "write")),
) -> SysPermissionResponse:
    operator = str(current_user.id)
    result = PermissionService(db).create(payload, operator)
    AuditLogService(db).write(
        user_id=current_user.id,
        username=current_user.username,
        action="CREATE",
        resource_type="SysPermission",
        resource_id=result.id,
        detail={"code": result.code, "action": result.action},
        ip_address=request.client.host if request.client else None,
    )
    _commit_write(db)
    return result


@router.get(
    "/permissions/{perm_id}",
    response_model=SysPermissionResponse,
    summary="Get permission",
)
def get_permission(
    perm_id: int,
    db: Session = Depends(get_db),
    _: SysUser = Depends(require_permission("/system/permissions", "read")),
) -> SysPermissionResponse:
    return PermissionService(db).get(perm_id)


@router.put(
    "/permissions/{perm_id}",
    response_model=SysPermissionResponse,
    summary="Update permission",
)
def update_permission(
    perm_id: int,
    payload: SysPermissionUpdate,
    request: Request,
    db: Session = Depends(get_db),
    current_user: SysUser = Depends(require_permission("/system/permissions", "write")),
) -> SysPermissionResponse:
    operator = str(current_user.id)
    result = PermissionService(db).update(perm_id, payload, operator)
    AuditLogService(db).write(
        user_id=current_user.id,
        username=current_user.username,
        action="UPDATE",
        resource_type="SysPermission",
        resource_id=perm_id,
        detail=payload.model_dump(exclude_unset=True),
        ip_address=request.client.host if request.client else None,
    )
    _commit_write(db)
    return result


@router.delete(
    "/permissions/{perm_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete permission",
)
def delete_permission(
    perm_id: int,
    request: Request,
    db: Session = Depends(get_db),
    current_user: SysUser = Depends(require_permission("/system/permissions", "delete")),
) -> None:
    PermissionService(db).delete(perm_id, str(current_user.id))
    AuditLogService(db).write(
        user_id=current_user.id,
        username=current_user.username,
        action="DELETE",
        resource_type="SysPermission",
        resource_id=perm_id,
        ip_address=request.client.host if request.client else None,
    )
    _commit_write(db)


@router.get(
    "/roles",
    response_model=PageResult[SysRoleResponse],
    summary="List roles",
)
def list_roles(
    keyword: str | None = Query(None, description="Role name or code keyword"),
    is_active: bool | None = Query(None),
    page: int = Query(1, ge=1),
    size: int = Query(20, ge=1, le=500),
    db: Session = Depends(get_db),
    _: SysUser = Depends(require_permission("/system/roles", "read")),
) -> PageResult[SysRoleResponse]:
    return RoleService(db).list(keyword=keyword, is_active=is_active, page=page, size=size)


@router.post(
    "/roles",
    response_model=SysRoleDetail,
    status_code=status.HTTP_201_CREATED,
    summary="Create role",
)
def create_role(
    payload: SysRoleCreate,
    request: Request,
    db: Session = Depends(get_db),
    current_user: SysUser = Depends(require_permission("/system/roles", "write")),
) -> SysRoleDetail:
    operator = str(current_user.id)
    result = RoleService(db).create(payload, operator)
    AuditLogService(db).write(
        user_id=current_user.id,
        username=current_user.username,
        action="CREATE",
        resource_type="SysRole",
        resource_id=result.id,
        detail={"name": result.name, "code": result.code},
        ip_address=request.client.host if request.client else None,
    )
    _commit_write(db)
    return result


@router.get(
    "/roles/{role_id}",
    response_model=SysRoleDetail,
    summary="Get role",
)
def get_role(
    role_id: int,
    db: Session = Depends(get_db),
    _: SysUser = Depends(require_permission("/system/roles", "read")),
) -> SysRoleDetail:
    return RoleService(db).get(role_id)


@router.put(
    "/roles/{role_id}",
    response_model=SysRoleDetail,
    summary="Update role",
    description="permission_ids replaces the full permission binding set when provided.",
)
def update_role(
    role_id: int,
    payload: SysRoleUpdate,
    request: Request,
    db: Session = Depends(get_db),
    current_user: SysUser = Depends(require_permission("/system/roles", "write")),
) -> SysRoleDetail:
    operator = str(current_user.id)
    result = RoleService(db).update(role_id, payload, operator)
    AuditLogService(db).write(
        user_id=current_user.id,
        username=current_user.username,
        action="UPDATE",
        resource_type="SysRole",
        resource_id=role_id,
        detail=payload.model_dump(exclude_unset=True, exclude={"permission_ids"}),
        ip_address=request.client.host if request.client else None,
    )
    _commit_write(db)
    return result


@router.delete(
    "/roles/{role_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete role",
)
def delete_role(
    role_id: int,
    request: Request,
    db: Session = Depends(get_db),
    current_user: SysUser = Depends(require_permission("/system/roles", "delete")),
) -> None:
    RoleService(db).delete(role_id, str(current_user.id))
    AuditLogService(db).write(
        user_id=current_user.id,
        username=current_user.username,
        action="DELETE",
        resource_type="SysRole",
        resource_id=role_id,
        ip_address=request.client.host if request.client else None,
    )
    _commit_write(db)


@router.get(
    "/users",
    response_model=PageResult[SysUserResponse],
    summary="List users",
)
def list_users(
    keyword: str | None = Query(None, description="Username, real name, or phone keyword"),
    is_active: bool | None = Query(None),
    department_id: int | None = Query(None, description="Department ID"),
    page: int = Query(1, ge=1),
    size: int = Query(20, ge=1, le=500),
    db: Session = Depends(get_db),
    _: SysUser = Depends(require_permission("/system/users", "read")),
) -> PageResult[SysUserResponse]:
    return UserService(db).list(
        keyword=keyword,
        is_active=is_active,
        department_id=department_id,
        page=page,
        size=size,
    )


@router.post(
    "/users",
    response_model=SysUserDetail,
    status_code=status.HTTP_201_CREATED,
    summary="Create user",
)
def create_user(
    payload: SysUserCreate,
    request: Request,
    db: Session = Depends(get_db),
    current_user: SysUser = Depends(require_permission("/system/users", "write")),
) -> SysUserDetail:
    operator = str(current_user.id)
    result = UserService(db).create(payload, operator)
    AuditLogService(db).write(
        user_id=current_user.id,
        username=current_user.username,
        action="CREATE",
        resource_type="SysUser",
        resource_id=result.id,
        detail={"username": result.username, "real_name": result.real_name},
        ip_address=request.client.host if request.client else None,
    )
    _commit_write(db)
    return result


@router.get(
    "/users/{user_id}",
    response_model=SysUserDetail,
    summary="Get user",
)
def get_user(
    user_id: int,
    db: Session = Depends(get_db),
    _: SysUser = Depends(require_permission("/system/users", "read")),
) -> SysUserDetail:
    return UserService(db).get(user_id)


@router.put(
    "/users/{user_id}",
    response_model=SysUserDetail,
    summary="Update user",
    description="role_ids replaces the full role binding set when provided.",
)
def update_user(
    user_id: int,
    payload: SysUserUpdate,
    request: Request,
    db: Session = Depends(get_db),
    current_user: SysUser = Depends(require_permission("/system/users", "write")),
) -> SysUserDetail:
    operator = str(current_user.id)
    result = UserService(db).update(user_id, payload, operator)
    AuditLogService(db).write(
        user_id=current_user.id,
        username=current_user.username,
        action="UPDATE",
        resource_type="SysUser",
        resource_id=user_id,
        detail=payload.model_dump(exclude_unset=True, exclude={"role_ids"}),
        ip_address=request.client.host if request.client else None,
    )
    _commit_write(db)
    return result


@router.delete(
    "/users/{user_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete user",
)
def delete_user(
    user_id: int,
    request: Request,
    db: Session = Depends(get_db),
    current_user: SysUser = Depends(require_permission("/system/users", "delete")),
) -> None:
    UserService(db).delete(user_id, str(current_user.id))
    AuditLogService(db).write(
        user_id=current_user.id,
        username=current_user.username,
        action="DELETE",
        resource_type="SysUser",
        resource_id=user_id,
        ip_address=request.client.host if request.client else None,
    )
    _commit_write(db)


@router.post(
    "/users/{user_id}/reset-password",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Reset user password",
    description="Admin resets a user password directly.",
)
def reset_user_password(
    user_id: int,
    body: SysUserAdminResetPassword,
    request: Request,
    db: Session = Depends(get_db),
    current_user: SysUser = Depends(require_permission("/system/users", "admin")),
) -> None:
    UserService(db).reset_password(user_id, body.new_password, str(current_user.id))
    AuditLogService(db).write(
        user_id=current_user.id,
        username=current_user.username,
        action="RESET_PASSWORD",
        resource_type="SysUser",
        resource_id=user_id,
        ip_address=request.client.host if request.client else None,
    )
    _commit_write(db)


@router.post(
    "/me/change-password",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Change current user password",
    description="Require current password and matching new password confirmation.",
)
def change_my_password(
    payload: SysUserChangePassword,
    request: Request,
    db: Session = Depends(get_db),
    current_user: SysUser = Depends(get_current_active_user),
) -> None:
    operator = str(current_user.id)
    UserService(db).change_password(current_user.id, payload, operator)
    AuditLogService(db).write(
        user_id=current_user.id,
        username=current_user.username,
        action="CHANGE_PASSWORD",
        resource_type="SysUser",
        resource_id=current_user.id,
        ip_address=request.client.host if request.client else None,
    )
    _commit_write(db)


@router.get(
    "/audit-logs",
    response_model=PageResult[SysAuditLogResponse],
    summary="List audit logs",
    description="Filter by user, action, resource type, and time range with pagination.",
)
def list_audit_logs(
    user_id: int | None = Query(None, description="Operator user ID"),
    action: str | None = Query(None, description="Action such as CREATE, UPDATE, DELETE"),
    resource_type: str | None = Query(None, description="Resource type such as SysUser or SysRole"),
    start_time: datetime | None = Query(None, description="Start time in ISO 8601"),
    end_time: datetime | None = Query(None, description="End time in ISO 8601"),
    page: int = Query(1, ge=1),
    size: int = Query(20, ge=1, le=500),
    db: Session = Depends(get_db),
    _: SysUser = Depends(require_permission("/system/audit-logs", "read")),
) -> PageResult[SysAuditLogResponse]:
    return AuditLogService(db).list(
        user_id=user_id,
        action=action,
        resource_type=resource_type,
        start_time=start_time,
        end_time=end_time,
        page=page,
        size=size,
    )

File: frontend/src/api/system.ts
import request from '@/utils/request'
import type { AxiosResponse } from 'axios'

// ═══════════════════════════════════════════════════════════════
// 通用类型
// ═══════════════════════════════════════════════════════════════

export interface PageResult<T> {
  items: T[]
  total: number
  page: number
  size: number
  pages: number
}

export interface PageQuery {
  keyword?: string
  page?: number
  size?: number
}

// ═══════════════════════════════════════════════════════════════
// 一、部门 Department
// ═══════════════════════════════════════════════════════════════

export interface Department {
  id: number
  name: string
  code: string
  parent_id: number | null
  sort_order: number
  is_active: boolean
  created_at: string
  updated_at: string
  created_by: string | null
  updated_by: string | null
}

export interface DeptQuery extends PageQuery {
  is_active?: boolean
}

export interface DeptCreate {
  name: string
  code: string
  parent_id?: number | null
  sort_order?: number
  is_active?: boolean
}

export interface DeptUpdate {
  name?: string
  code?: string
  parent_id?: number | null
  sort_order?: number
  is_active?: boolean
}

export const deptApi = {
  list: (params: DeptQuery): Promise<AxiosResponse<PageResult<Department>>> =>
    request.get('/system/departments', { params }),

  create: (data: DeptCreate): Promise<AxiosResponse<Department>> =>
    request.post('/system/departments', data),

  detail: (id: number): Promise<AxiosResponse<Department>> =>
    request.get(`/system/departments/${id}`),

  update: (id: number, data: DeptUpdate): Promise<AxiosResponse<Department>> =>
    request.put(`/system/departments/${id}`, data),

  remove: (id: number): Promise<AxiosResponse<void>> =>
    request.delete(`/system/departments/${id}`),
}

// ═══════════════════════════════════════════════════════════════
// 二、权限 Permission
// ═══════════════════════════════════════════════════════════════

export type PermissionAction = 'read' | 'write' | 'delete' | 'admin'

export interface Permission {
  id: number
  name: string
  code: string
  resource: string
  action: PermissionAction
  description: string | null
  parent_id: number | null
  created_at: string
  updated_at: string
}

export interface PermQuery extends PageQuery {
  action?: PermissionAction
}

export interface PermCreate {
  name: string
  code: string
  resource: string
  action: PermissionAction
  description?: string | null
  parent_id?: number | null
}

export type PermUpdate = Partial<PermCreate>

export const permApi = {
  list: (params: PermQuery): Promise<AxiosResponse<PageResult<Permission>>> =>
    request.get('/system/permissions', { params }),

  create: (data: PermCreate): Promise<AxiosResponse<Permission>> =>
    request.post('/system/permissions', data),

  detail: (id: number): Promise<AxiosResponse<Permission>> =>
    request.get(`/system/permissions/${id}`),

  update: (id: number, data: PermUpdate): Promise<AxiosResponse<Permission>> =>
    request.put(`/system/permissions/${id}`, data),

  remove: (id: number): Promise<AxiosResponse<void>> =>
    request.delete(`/system/permissions/${id}`),
}

// ═══════════════════════════════════════════════════════════════
// 三、角色 Role
// ═══════════════════════════════════════════════════════════════

export interface RoleBase {
  id: number
  name: string
  code: string
  description: string | null
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface RoleDetail extends RoleBase {
  permissions: Pick<Permission, 'id' | 'name' | 'code' | 'action'>[]
}

export interface RoleQuery extends PageQuery {
  is_active?: boolean
}

export interface RoleCreate {
  name: string
  code: string
  description?: string | null
  is_active?: boolean
  permission_ids?: number[]
}

export interface RoleUpdate {
  name?: string
  code?: string
  description?: string | null
  is_active?: boolean
  permission_ids?: number[]
}

export const roleApi = {
  list: (params: RoleQuery): Promise<AxiosResponse<PageResult<RoleBase>>> =>
    request.get('/system/roles', { params }),

  create: (data: RoleCreate): Promise<AxiosResponse<RoleDetail>> =>
    request.post('/system/roles', data),

  detail: (id: number): Promise<AxiosResponse<RoleDetail>> =>
    request.get(`/system/roles/${id}`),

  update: (id: number, data: RoleUpdate): Promise<AxiosResponse<RoleDetail>> =>
    request.put(`/system/roles/${id}`, data),

  remove: (id: number): Promise<AxiosResponse<void>> =>
    request.delete(`/system/roles/${id}`),
}

// ═══════════════════════════════════════════════════════════════
// 四、用户 User
// ═══════════════════════════════════════════════════════════════

export interface UserBase {
  id: number
  username: string
  real_name: string
  email: string | null
  phone: string | null
  is_active: boolean
  department_id: number | null
  created_at: string
  updated_at: string
  created_by: string | null
  updated_by: string | null
}

export interface UserDetail extends UserBase {
  department: Pick<Department, 'id' | 'name' | 'code'> | null
  roles: Pick<RoleBase, 'id' | 'name' | 'code' | 'is_active'>[]
}

export interface UserQuery extends PageQuery {
  is_active?: boolean
  department_id?: number
}

export interface UserCreate {
  username: string
  password: string
  real_name: string
  email?: string | null
  phone?: string | null
  is_active?: boolean
  department_id?: number | null
  role_ids?: number[]
}

export interface UserUpdate {
  real_name?: string
  email?: string | null
  phone?: string | null
  is_active?: boolean
  department_id?: number | null
  role_ids?: number[]
}

export interface ResetPasswordPayload {
  new_password: string
}

export interface ChangePasswordPayload {
  old_password: string
  new_password: string
  confirm_password: string
}

export const userApi = {
  list: (params: UserQuery): Promise<AxiosResponse<PageResult<UserBase>>> =>
    request.get('/system/users', { params }),

  create: (data: UserCreate): Promise<AxiosResponse<UserDetail>> =>
    request.post('/system/users', data),

  detail: (id: number): Promise<AxiosResponse<UserDetail>> =>
    request.get(`/system/users/${id}`),

  update: (id: number, data: UserUpdate): Promise<AxiosResponse<UserDetail>> =>
    request.put(`/system/users/${id}`, data),

  remove: (id: number): Promise<AxiosResponse<void>> =>
    request.delete(`/system/users/${id}`),

  resetPassword: (id: number, data: ResetPasswordPayload): Promise<AxiosResponse<void>> =>
    request.post(`/system/users/${id}/reset-password`, data),

  changePassword: (data: ChangePasswordPayload): Promise<AxiosResponse<void>> =>
    request.post('/system/me/change-password', data),
}

// ═══════════════════════════════════════════════════════════════
// 五、审计日志 AuditLog（只读）
// ═══════════════════════════════════════════════════════════════

export type AuditAction = string
export type AuditResourceType = string

export interface AuditLog {
  id: number
  user_id: number
  username: string
  action: AuditAction
  resource_type: AuditResourceType
  resource_id: string
  detail: Record<string, unknown> | null
  ip_address: string | null
  user_agent: string | null
  created_at: string
}

export interface AuditQuery {
  user_id?: number
  action?: AuditAction
  resource_type?: AuditResourceType
  start_time?: string
  end_time?: string
  page?: number
  size?: number
}

export const auditApi = {
  list: (params: AuditQuery): Promise<AxiosResponse<PageResult<AuditLog>>> =>
    request.get('/system/audit-logs', { params }),
}
